### PR TITLE
use mallocMC for species storage

### PR DESCRIPTION
**Pull request by psychocoderHPC**
This pull request use mallocMC to manage the shared species memory.

- add `mallocMC` to `CMakeLists.txt`
- add PointerWrapper class `Pointer<>`
- delete old Heap- and RingBuffer/Boxes
- **remove the possibility that species device memory can directly copied to host memory**

Tests:
- [x] KHI heating comparison to current dev
- [x] add more documentation

~~Do not merge before~~ #717

### CODE REVIEW

File: src/picongpu/CMakeLists.txt [hunk 1]
```diff
@@ -253,6 +252,12 @@ if( (Boost_VERSION EQUAL 105500) AND
       "${CUDA_NVCC_FLAGS} \"-DBOOST_NOINLINE=__attribute__((noinline))\" ")
 endif()
 
+################################################################################
+# Find mallocMC
+###############################################################################
+
+find_package(mallocMC 2.0.1 REQUIRED)
```
Comment (ax3l) [inline src/picongpu/CMakeLists.txt]: for now: `+EXACT`

File: src/libPMacc/include/particles/memory/boxes/ParticlesBox.hpp [hunk 2]
```diff
@@ -64,7 +67,16 @@ class ParticlesBox
      */
     HDINLINE FRAME &getEmptyFrame()
     {
-        return data.pop();
+        FrameType* tmp = new FrameType;
+
+        if (tmp != NULL)
+        {
+            /* delete all particles we can not assume that new memory is zeroed */
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/memory/boxes/ParticlesBox.hpp]: better:
delete all particles **since** we can not assume that **newly allocated** memory contains **zeros**

Do you realy *delete* particles here or do you set *some attribut* of them to zero?

File: src/libPMacc/include/particles/memory/boxes/ParticlesBox.hpp [hunk 3]
```diff
@@ -250,49 +230,36 @@ class ParticlesBox
...
+        if (last.isValid())
         {
-            //prev_id is Valid
-            next[prev_id] = INV_IDX; //clear next of previous frame
-            *lastFrameIdx = prev_id; //set new last particle
-            removeFrame(data[last_id]);
-            return true;
+
+            FramePtr prev(last->previousFrame);
+            //            last->previousFrame = FramePtr(); //delete previous frame of the frame which we remove
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/memory/boxes/ParticlesBox.hpp]: do you **delete** the previous frame here?
Or do you just set the pointer to the previous frame?

In this case:
after deleting frame, set pointer to previous frame

File: src/libPMacc/include/particles/memory/boxes/ParticlesBox.hpp [hunk 4]
```diff
@@ -250,49 +230,36 @@ class ParticlesBox
...
+
+            if (prev.isValid())
+            {
+                prev->nextFrame = FramePtr(); //clear next of previous frame
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/memory/boxes/ParticlesBox.hpp]: what do you mean by `clear next`?

File: src/libPMacc/include/particles/memory/dataTypes/Pointer.hpp [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+class Pointer
+{
+public:
+
+    typedef T_Type type;
+    typedef type* PtrType;
+
+    /** default constructor
+     *
+     * the default Pointer points to invalid memory
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/memory/dataTypes/Pointer.hpp]: `Pointer` -> `pointer`

File: src/libPMacc/include/particles/memory/dataTypes/Pointer.hpp [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        return *ptr;
+    }
+
+    /** access member*/
+    HDINLINE PtrType operator->()
+    {
+        return ptr;
+    }
+
+    /** compare if two pointer points to the same memory address*/
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/memory/dataTypes/Pointer.hpp]: compare if *the* two **pointers point** to the same memory address

File: src/libPMacc/include/particles/memory/dataTypes/Pointer.hpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    HDINLINE bool operator==(const Pointer<type>& other) const
+    {
+        return ptr == other.ptr;
+    }
+
+    /** check if the memory address of two pointer is different*/
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/memory/dataTypes/Pointer.hpp]: check if the memory address of *the* two **pointers are** different

File: src/libPMacc/include/particles/memory/dataTypes/Pointer.hpp [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    HDINLINE bool operator!=(const Pointer<type>& other) const
+    {
+        return ptr != other.ptr;
+    }
+
+    /** check if the pointed memory is has valid address
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/memory/dataTypes/Pointer.hpp]: check if the memory pointed to ~~is~~ has *a* valid address

File: src/libPMacc/include/particles/memory/frames/Frame.hpp [hunk 9]
```diff
@@ -149,6 +154,12 @@ public InheritLinearly<typename T_ParticleDescription::MethodsList>,
         return std::string(boost::mpl::c_str<Name>::value);
     }
 
+    /* \todo find a generic solution to add `previousFrame` and `nextFrame`
+     * only if we use this frame in a double linked list
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/memory/frames/Frame.hpp]: Do you mean:

**use this part of the code** only if we use this frame in a double linked list

File: src/picongpu/include/simulationControl/MySimulation.hpp [hunk 10]
```diff
@@ -455,15 +457,15 @@ class MySimulation : public SimulationHelper<simDim>
...
-        // global size must a devisor of supercell size
-        // note: this is redundant, while using the local condition below
-
-        assert(globalGridSize[i] % MappingDesc::SuperCellSize::toRT()[i] == 0);
-        // local size must a devisor of supercell size
-        assert(gridSizeLocal[i] % MappingDesc::SuperCellSize::toRT()[i] == 0);
-        // local size must be at least 3 supercells (1x core + 2x border)
-        // note: size of border = guard_size (in supercells)
-        // \todo we have to add the guard_x/y/z for modified supercells here
+            // global size must a devisor of supercell size
```
Comment (PrometheusPi) [inline src/picongpu/include/simulationControl/MySimulation.hpp]: global size must **be** a devisor of supercell size

File: src/picongpu/include/simulationControl/MySimulation.hpp [hunk 11]
```diff
@@ -455,15 +457,15 @@ class MySimulation : public SimulationHelper<simDim>
...
+            // note: this is redundant, while using the local condition below
+
+            assert(globalGridSize[i] % MappingDesc::SuperCellSize::toRT()[i] == 0);
+            // local size must a devisor of supercell size
```
Comment (PrometheusPi) [inline src/picongpu/include/simulationControl/MySimulation.hpp]: as above

File: src/picongpu/main.cu [hunk 12]
```diff
@@ -29,6 +29,49 @@
  */
 
 
+//include the Heap with the arguments given in the config
```
Comment (PrometheusPi) [inline src/picongpu/main.cu]: `Heap`-> `heap`
What do you mean by this comment?

File: src/libPMacc/include/particles/memory/frames/Frame.hpp (hunk 9 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/memory/frames/Frame.hpp]: Yes, we need this attributes only in a double linked list.

File: src/picongpu/CMakeLists.txt [hunk 13]
```diff
@@ -253,6 +252,12 @@ if( (Boost_VERSION EQUAL 105500) AND
       "${CUDA_NVCC_FLAGS} \"-DBOOST_NOINLINE=__attribute__((noinline))\" ")
 endif()
 
+################################################################################
```
Comment (ax3l) [inline src/picongpu/CMakeLists.txt]: ok for now, but as talked offline: we will simply add a include in the final version to reduce install dependencies.

**Comment by ax3l** [comment]
looks good :+1:

**Comment by psychocoderHPC** [comment]
This version needs a small refactoring against the current mallocMC release. 
I am back tomorrow and will do this.

**Comment by ax3l** [comment]
fantastico! :sparkles:

**Comment by psychocoderHPC** [comment]
After our compile suite is updated it is ready to merge.

**Comment by psychocoderHPC** [comment]
electron/positron heating
heating

**Comment by ax3l** [comment]
> After our compile suite is updated it is ready to merge.

I will update it asap - nevertheless, I think we should
- [ ] import mallocMC via subtree
before we can merge :)

File: src/picongpu/CMakeLists.txt (hunk 13 above)
Comment (ax3l) [inline src/picongpu/CMakeLists.txt]: opened #717

**Comment by psychocoderHPC** [comment]
rebased against #717

File: src/picongpu/CMakeLists.txt [hunk 14]
```diff
@@ -39,7 +39,6 @@ set(CMAKE_PREFIX_PATH "/usr/lib/x86_64-linux-gnu/" "$ENV{MPI_ROOT}"
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH}
     ${CMAKE_CURRENT_SOURCE_DIR}/../../thirdParty/cmake-modules/)
 
-
```
Comment (ax3l) [inline src/picongpu/CMakeLists.txt]: pls revert the changes in this file

File: src/libPMacc/include/particles/memory/boxes/ParticlesBox.hpp [hunk 15]
```diff
@@ -81,59 +92,35 @@ class ParticlesBox
      *
      * @param frame FRAME to remove
      */
-    HDINLINE void removeFrame(FRAME &frame)
+    DINLINE void removeFrame(const FRAME &frame)
```
Comment (ax3l) [inline src/libPMacc/include/particles/memory/boxes/ParticlesBox.hpp]: is free-ing a `const ref` a good idea?

**Comment by ax3l** [comment]
It would be cool if we could add a plain benchmark "before" - "after" mallocMC (without plugins) to just document the performance stayed the same.
(last time we did that check we found some bugs that were then [fixed in mallocMC](https://github.com/ComputationalRadiationPhysics/mallocMC/pull/84/files?short_path=4ac32a7#diff-4ac32a78649ca5bdd8e0ba38b7006a1e))

**Comment by psychocoderHPC** [comment]
Runtime KHI 3d Esirkepov [in seconds]:

|   |  init | simulation |
|---|---|---|
| dev  | 26   | 1575 |  
| branch  | 10   | 1569   |

**Comment by ax3l** [comment]
well done, good work! :+1:

**Comment by PrometheusPi** [comment]
:+1: cool
